### PR TITLE
Tests for metrics calculation + Torchmetrics for IoU

### PR DESCRIPTION
**Pull request by mpelchat04**
Adds tests for metrics calculation (classification report and IoU).  

IoU now relies on Torchmetrics for calculation.   

Classification report still uses scikit-learn because torchmetrics does not support ignore_index == -1.

### CODE REVIEW

File: tests/utils/test_metrics.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+from utils.metrics import create_metrics_dict, report_classification, iou
+import torch
+import pytest
+
+# Test arrays: [bs=2, h=2, w,2]
```
Comment (CharlesAuthier) [inline tests/utils/test_metrics.py:5]: do you mean w=2?

File: tests/utils/test_metrics.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        """Evaluate report classification. 
+        Multiclass, without ignore_index in array."""
+        t = init_tensors()
+        metrics_dict = create_metrics_dict(3)
+        metrics_dict = report_classification(t['pred_multi'], 
+                                             t['lbl_multi'], 
+                                             batch_size=2, 
+                                             metrics_dict=metrics_dict,
+                                             ignore_index=-1)
+        assert "{:.6f}".format(metrics_dict['precision'].val) == "0.327083"
```
Comment (CharlesAuthier) [inline tests/utils/test_metrics.py:50]: Do we want a log of this in case of test on others machines like HPC?

**Review by CharlesAuthier — APPROVED** [review]

File: tests/utils/test_metrics.py (hunk 2 above)
Comment (remtav) [reply tests/utils/test_metrics.py:50]: I don't see the necessity of logs for unit tests. Pytest does output a pretty detailed log if 
tests fail...

File: tests/utils/test_metrics.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        t = init_tensors()
+        metrics_dict = create_metrics_dict(3)
+        metrics_dict = report_classification(t['pred_multi'], 
+                                             t['lbl_multi_dc'], 
+                                             batch_size=2, 
+                                             metrics_dict=metrics_dict,
+                                             ignore_index=-1)
+        assert "{:.6f}".format(metrics_dict['precision'].val) == "0.297619"
+        assert "{:.6f}".format(metrics_dict['recall'].val) == "0.285714"
+        assert "{:.6f}".format(metrics_dict['fscore'].val) == "0.283163"
```
Comment (remtav) [inline tests/utils/test_metrics.py:66]: If you'd like to one step further in this verification, you could define some functions for precision, recall, fscore and iou from numpy directly and compare results from torchmetrics with these simple functions. This assertion is enough in my opinion though. The only downside is that it doesn't "prove" that a metric is being calculated the right way, it just proves that it's constant.

File: utils/metrics.py [hunk 4]
```diff
@@ -1,17 +1,25 @@
...
+from math import sqrt
+from torch import IntTensor
+from torchmetrics import JaccardIndex
 
 min_val = 1e-6
-def create_metrics_dict(num_classes):
-    num_classes = num_classes if num_classes == 1 else num_classes + 1
+def create_metrics_dict(num_classes, ignore_index=None):
+
+    num_classes = num_classes + 1 if num_classes == 1 else num_classes
```
Comment (remtav) [inline utils/metrics.py:12]: That's a pretty darn subtle bug :(
High five to the digital accelerator folks...

**Review by remtav — APPROVED** [review]

File: tests/utils/test_metrics.py (hunk 1 above)
Comment (mpelchat04) [reply tests/utils/test_metrics.py:5]: Damn it! En effet.

File: tests/utils/test_metrics.py (hunk 3 above)
Comment (mpelchat04) [reply tests/utils/test_metrics.py:66]: Right. I took the easy path of calculating the values in an excel sheet, to validate them.

File: utils/metrics.py (hunk 4 above)
Comment (mpelchat04) [reply utils/metrics.py:12]: Yup. But with the tests, we'll see these errors now!

**Review by CharlesAuthier — APPROVED** [review]

File: tests/utils/test_metrics.py (hunk 3 above)
Comment (remtav) [reply tests/utils/test_metrics.py:66]: That's perfect. Thanks for validating!

**Review by remtav — APPROVED** [review]
